homeassistant/components/denonavr/config_flow.py

```diff
@@ -0,0 +1,230 @@
+
+DOMAIN = "denonavr"
+
+SUPPORTED_MANUFACTURERS = ["Denon", "DENON", "Marantz"]
+
+CONF_SHOW_ALL_SOURCES = "show_all_sources"
+CONF_ZONE2 = "zone2"
+CONF_ZONE3 = "zone3"
+CONF_TYPE = "type"
+CONF_MODEL = "model"
+CONF_MANUFACTURER = "manufacturer"
+
+DEFAULT_SHOW_SOURCES = False
+DEFAULT_TIMEOUT = 5
+DEFAULT_ZONE2 = False
+DEFAULT_ZONE3 = False
+
+CONFIG_SCHEMA = vol.Schema({vol.Optional(CONF_HOST): str})
+
+SETTINGS_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_SHOW_ALL_SOURCES, default=DEFAULT_SHOW_SOURCES): bool,
+        vol.Optional(CONF_ZONE2, default=DEFAULT_ZONE2): bool,
+        vol.Optional(CONF_ZONE3, default=DEFAULT_ZONE3): bool,
+    }
+)
+
+
+class DenonAvrFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a Denon AVR config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    def __init__(self):
+        """Initialize the Denon AVR flow."""
+        self.host = None
+        self.serial_number = None
+        self.model_name = None
+        self.timeout = DEFAULT_TIMEOUT
+        self.show_all_sources = DEFAULT_SHOW_SOURCES
+        self.zone2 = DEFAULT_ZONE2
+        self.zone3 = DEFAULT_ZONE3
+        self.d_receivers = []
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initialized by the user."""
+        errors = {}
+        if user_input is not None:
+            # check if IP address is set manually
+            host = user_input.get(CONF_HOST)
+            if host:
+                self.host = host
+                return await self.async_step_settings()
+
+            # discovery using denonavr library
+            self.d_receivers = await self.hass.async_add_executor_job(denonavr.discover)
+            # More than one receiver could be discovered by that method
+            if len(self.d_receivers) == 1:
+                self.host = self.d_receivers[0]["host"]
+                return await self.async_step_settings()
+            if len(self.d_receivers) > 1:
+                # show selection form
+                return await self.async_step_select()
+
+            errors["base"] = "discovery_error"
+
+        return self.async_show_form(
+            step_id="user", data_schema=CONFIG_SCHEMA, errors=errors
+        )
+
+    async def async_step_select(self, user_input=None):
+        """Handle multiple receivers found."""
+        errors = {}
+        if user_input is not None:
+            self.host = user_input["select_host"]
+            return await self.async_step_settings()
+
+        select_scheme = vol.Schema(
+            {
+                vol.Required("select_host"): vol.In(
+                    [d_receiver["host"] for d_receiver in self.d_receivers]
+                )
+            }
+        )
+
+        return self.async_show_form(
+            step_id="select", data_schema=select_scheme, errors=errors
+        )
+
+    async def async_step_settings(self, user_input=None):
```

Comment:
Optional settings should not be part of the main flow but an options flow. If you don't want to add the options flow in this PR then remove these options for now from the flow.

Reply:
I am not going to implement Options Flow. I have no idea how that works and frankly I am kind of done with this PR.
I can't remove those options because users will not be able to setup there zones and that will make a lot of users really angry...

Reply:
done